src/split_reduce.cpp

```diff
@@ -0,0 +1,208 @@
+#include <migraphx/register_op.hpp>
+#include <migraphx/functional.hpp>
+#include <migraphx/algorithm.hpp>
+#include <migraphx/param_utils.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+struct split_fused_reduce
+{
+    std::vector<std::int64_t> axes{};
+    std::string assign = "assign_none";
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.axes, "axes"), f(self.assign, "assign"));
+    }
+
+    value attributes() const { return {{"prefill", 0}}; }
+
+    shape compute_shape(const std::vector<shape>& inputs, std::vector<module_ref> mods) const
+    {
+        if(mods.size() != 1)
+            MIGRAPHX_THROW("should have one submodule.");
+        const auto* sm = mods.front();
+        if(sm->get_output_shapes().size() != 1)
+            MIGRAPHX_THROW("Only one output supported");
+        auto names = sm->get_parameter_names();
+        check_shapes{inputs, *this}.has(names.size()).same_ndims();
+        std::sort(names.begin(), names.end());
+        auto shapes = sm->get_parameter_shapes();
+        // Check dimension matches for each input
+        if(not equal(names, inputs, [&](const auto& name, const auto& input) {
+               return shapes.at(name).lens() == input.lens();
+           }))
+            MIGRAPHX_THROW("Dimenstion does not match the submodule.");
+        const auto& s = inputs.at(0);
+        auto lens     = s.lens();
+        if(lens != sm->get_output_shapes().front().lens())
+        {
+            for(const auto& axis : axes)
+            {
+                lens[axis] = 1;
+            }
+        }
+
+        return shape::from_permutation(
+            sm->get_output_shapes().front().type(), lens, find_permutation(inputs));
+    }
+
+    std::string name() const { return "split_fused_reduce"; }
+};
+MIGRAPHX_REGISTER_OP(split_fused_reduce);
+
+static bool is_reduce(const instruction& ins) { return contains(ins.name(), "reduce"); }
+
+static std::vector<instruction_ref> find_split(const_module_ref rm)
+{
+    std::vector<instruction_ref> result;
+    auto reduce_ins = std::find_if(rm->begin(), rm->end(), &is_reduce);
+    if(reduce_ins == rm->end())
+        return result;
+    // Bail if there is more than one reduce for now
+    if(std::any_of(std::next(reduce_ins), rm->end(), &is_reduce))
+        return result;
+    // Only handle reduce_sum for now
+    if(reduce_ins->name() != "reduce_sum")
+        return result;
+    result.push_back(reduce_ins);
+    // TODO: Find instructions that are used again in the module
+    return result;
+}
+
+static std::vector<instruction_ref> get_alive(const_module_ref rm,
+                                              const std::vector<instruction_ref>& splits)
+{
+    std::vector<instruction_ref> result;
+    bool stop = false;
+    liveness(*rm, [&](auto ins, const auto& live_set) {
+        if(stop)
+            return;
+        if(not contains(splits, ins))
+            return;
```

Comment:
Why do you require a stop flag ? 
`if(not contains(splits,ins))` would take care of early return.

Reply:
We only want the liveness for the first split found. The `return` is like a `continue` in the loop and the `stop` flag is like a `break`.

Reply:
> and the stop flag is like a break. 

Yeah but i don't see need to have `stop` flag.

Reply:
Then how else can I `break` on the first split?

Reply:
> Then how else can I break on the first split? 

You don't need to break, `if(not contains(splits, ins))` would break it for you after the first split.

Reply:
Yea I do need a break. I only want the instructions live at the first split found(which is the last split instructions in the module due to how liveness traverses in reverse). I dont care whats live on any previous splits.

If I didnt break then I would be splitting the module at unnecessary points(also there would be duplicates for instructions live for both).

Right now we dont handle multiple outputs so there is only one split so it wouldn't make a difference, but the goal is to support multiple outputs which will require multiple split points.